### DISCUSSION TITLE
Handling object arrays as inputs for components

### DISCUSSION THREAD

**Discussion by TheMouster** [defx/synergy#6] [General]
Hi,

If I pass an object array in as an attribute value, how can I iterate the array in the component output?

Thanks

**Comment by TheMouster** [comment]
E.g. 
`<fms-faults list=‘[{“type”:”10”,”text”:”blah blah blah”},{“type”:”20”,”text”:”doh doh doh”},{“type”:”30”,”text”:”duh duh duh”}]'></fms-faults>`

**Reply by defx** [reply]
Synergy accepts rich data as properties rather than attributes. To understand why check https://developers.google.com/web/fundamentals/web-components/best-practices#aim-to-only-accept-rich-data-objects,-arrays-as-properties.

**Reply by defx** [reply]
You could rewrite your example above as follows:

```html 
  <body>
    <fms-faults></fms-faults>
    <script type="module">
      import { define } from "https://unpkg.com/synergy";

      define("fms-faults", ({}) => ({
        list: [
          { type: 10, text: "blah blah blah" },
          { type: 20, text: "doh doh doh" },
          { type: 30, text: "duh duh duh" },
        ],
      }), document.querySelector("template#fms-faults"));
    </script>
    <template id="fms-faults">
      <template each="item in list">
        <h3>{{ item.type }}</h3>
        <p>{{ item.text }}</p>
      </template>
    </template>
  </body>
```

**Reply by defx** [reply]
I'm not sure what you actually want to do in your template, but this shows you how to provide hard-coded data to your template using the model returned from your factory.

**Reply by defx** [reply]
Its also possible to pass rich data from one component down to another via attribute interpolation:

```html
  <body>
    <fms-foo></fms-foo>
    <script type="module">
      import { define } from "https://unpkg.com/synergy";

      define("fms-foo", ({}) => ({
        items: [
          { type: 10, text: "blah blah blah" },
          { type: 20, text: "doh doh doh" },
          { type: 30, text: "duh duh duh" },
        ],
      }), document.querySelector("template#fms-foo"));

      define("fms-faults", ({ list = [] }) => ({
        list,
      }), document.querySelector("template#fms-faults"));
    </script>
    <template id="fms-foo">
      <fms-faults list="{{ items }}"></fms-faults>
    </template>
    <template id="fms-faults">
      <template each="item in list">
        <h3>{{ item.type }}</h3>
        <p>{{ item.text }}</p>
      </template>
    </template>
  </body>
```

...under the hood, Synergy just checks to see whether the value being passed is non-primitive and, if it is, then it passes that value as a property rather than an attribute.

**Reply by defx** [reply]
If you wanted to load some rich data for your page and then provide that to your top-level component, you can do this in the same way that you would set a property on any other HTML element...

```html
  <body>
    <fms-faults id="myElement"></fms-faults>

    <script type="module">
      import { define } from "https://unpkg.com/synergy";

      define("fms-faults", ({ list = [] }) => ({
        list,
      }), document.querySelector("template#fms-faults"));
    </script>
    <template id="fms-faults">
      <template each="item in list">
        <h3>{{ item.type }}</h3>
        <p>{{ item.text }}</p>
      </template>
    </template>
    <script>
      document.getElementById("myElement").list = [
        { type: 10, text: "blah blah blah" },
        { type: 20, text: "doh doh doh" },
        { type: 30, text: "duh duh duh" },
      ];
    </script>
  </body>
```

**Reply by TheMouster** [reply]
Hi,

I couldn’t get either of these to work the way I wanted. I’ll return and give you a more detailed example and get your opinion on whether what I would like to have happen is possible. Unfortunately, like most developers, I’m running seriously behind on my current project :-D. Back to the grindstone.

Cheers
